kernel/kernel-api/src/test/scala/io/delta/kernel/internal/types/DataTypeJsonSerDeSuite.scala

```diff
@@ -245,6 +274,249 @@ object DataTypeJsonSerDeSuite {
+            new ArrayType(
+              new ArrayType(StringType.STRING, true), true),
+            new StructType()
+              .add("c1", StringType.STRING, false)
+              .add("c2", StringType.STRING, true), true))
+          .add("b2", LongType.LONG), true)
+        .add("a3", new ArrayType(
+          new MapType(
+            StringType.STRING,
+            new StructType()
+              .add("b1", DateType.DATE, false), false), false), true)
+    )
+  )
+
+  val SAMPLE_JSON_TO_TYPES_WITH_COLLATION = Seq(
+    (
+      structTypeJson(Seq(
+        structFieldJson("a1", "\"string\"", true,
+          metadataJson = Some(s"""{"$COLLATIONS_METADATA_KEY" : {"a1" : "ICU.UNICODE"}}""")),
+        structFieldJson("a2", "\"integer\"", false),
+        structFieldJson("a3", "\"string\"", false,
+          metadataJson = Some(s"""{"$COLLATIONS_METADATA_KEY" : {"a3" : "SPARK.UTF8_LCASE"}}""")),
+        structFieldJson("a4", "\"string\"", true))),
+      new StructType()
+        .add("a1", new StringType("ICU.UNICODE"), true)
+        .add("a2", IntegerType.INTEGER, false)
+        .add("a3", new StringType("SPARK.UTF8_LCASE"), false)
+        .add("a4", StringType.STRING, true)
+    ),
+    (
+      structTypeJson(Seq(
+        structFieldJson("a1", structTypeJson(Seq(
+          structFieldJson("b1", "\"string\"", true,
+            metadataJson = Some(
+              s"""{"$COLLATIONS_METADATA_KEY"
+                 | : {"b1" : "ICU.UNICODE"}}""".stripMargin)))), true),
+        structFieldJson("a2", structTypeJson(Seq(
+          structFieldJson("b1", arrayTypeJson("\"string\"", false), true,
+            metadataJson = Some(
+              s"""{"$COLLATIONS_METADATA_KEY"
+                 | : {"b1.element" : "SPARK.UTF8_LCASE"}}""".stripMargin)),
+          structFieldJson("b2", mapTypeJson("\"string\"", "\"string\"", true), false,
+            metadataJson = Some(
+              s"""{"$COLLATIONS_METADATA_KEY"
+                 | : {"b2.key" : "ICU.UNICODE_CI",
+                 |  "b2.value" : "SPARK.UTF8_LCASE"}}""".stripMargin)),
+          structFieldJson("b3", arrayTypeJson("\"string\"", false), true),
+          structFieldJson("b4", mapTypeJson("\"string\"", "\"string\"", false), false))), true),
+        structFieldJson("a3", structTypeJson(Seq(
+          structFieldJson("b1", "\"string\"", false),
+          structFieldJson("b2", arrayTypeJson("\"integer\"", false), true))), false,
+          metadataJson = Some(
+            s"""{"$COLLATIONS_METADATA_KEY"
+               | : {"b1" : "SPARK.UTF8_LCASE"}}""".stripMargin)))),
+      new StructType()
+        .add("a1", new StructType()
+          .add("b1", new StringType("ICU.UNICODE")), true)
+        .add("a2", new StructType()
+          .add("b1", new ArrayType(new StringType("SPARK.UTF8_LCASE"), false))
+          .add("b2", new MapType(
+            new StringType("ICU.UNICODE_CI"), new StringType("SPARK.UTF8_LCASE"), true), false)
```

Comment:
what about the case where Map/array element is a struct which has a string column with collation.